OAK-9918: DocumentNodeStore consistency check

mreutegg opened the pull request.

mreutegg committed:
OAK-9918: DocumentNodeStore consistency check

mreutegg committed:
OAK-9918: DocumentNodeStore consistency check

Turn numThreads system property into a proper option
Add some documentation to README.md

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java:
```diff
@@ -0,0 +1,159 @@
+        if (!executorService.awaitTermination(5, TimeUnit.MINUTES)) {
+            String msg = "Checks still not finished after the last one has been submitted 5 minutes ago";
+            results.put(() -> {
+                StringBuilder sb = new StringBuilder("{");
```
Can we please use JsonObject to create JSON?

reschke replied:
For better or worse, Oak already has it's own JSON library. IMHO that should be used.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java:
```diff
@@ -0,0 +1,159 @@
+
+        @Override
+        public String toJson() {
+            return "{\"orphaned\": \"" + orphaned + "\", \"revision\": \"" + revision + "\", \"missing\": \"" + missing + "\"}";
```
Same as above, use Json lib to create json.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/Progress.java:
```diff
@@ -0,0 +1,62 @@
+
+        @Override
+        public String toJson() {
+            StringBuilder sb = new StringBuilder("{");
```
Same as above, please use jsonobject

rishabhdaim replied:
```suggestion
            JsonObject json = new JsonObject();
        json.getProperties().put("time", nowAsISO8601());
        json.getProperties().put("info", msg);
        return json.toString();
```

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/ProgressWithETA.java:
```diff
@@ -0,0 +1,49 @@
+            @Override
+            public String toJson() {
+                StringBuilder sb = new StringBuilder("{");
+                sb.append("\"time\": \"").append(nowAsISO8601()).append('"');
```
Same as above, please use jsonobject.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/Summary.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public void end(@NotNull BlockingQueue<Result> results)
+            throws InterruptedException {
+        results.put(() -> "{\"summary\": \"Checked " + numDocuments + " documents in " + sw + ". Number of threads used: " + numThreads + "\"}");
```
Same as above

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/DocumentProcessor.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public interface DocumentProcessor {
+
+    void processDocument(@NotNull NodeDocument document,
```
Could you please add JavaDocs for this api.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java:
```diff
@@ -0,0 +1,159 @@
+        new ExecutorCloser(executorService, 5, TimeUnit.MINUTES).close();
+    }
+
+    private static final class CheckDocument implements Callable<Void> {
```
I would change to implement `Runnbale` since we are neither returning anything nor throwing any exceptions.

mreutegg replied:
Unfortunately that's not quite true. The task implementation puts a `Result` in the `results` queue, which may throw an `InterruptedException`. I'll leave this as is.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java:
```diff
@@ -0,0 +1,159 @@
+            return null;
+        }
+
+        private Path assertAncestorsExist(Path path) {
```
```suggestion
        private Optional<Path> assertAncestorsExist(Path path) {
        if (path == null) {
            return Optional.empty();
        }
        NodeState state = ns.getRoot();
        Path p = Path.ROOT;
        for (String name : path.elements()) {
            p = new Path(p, name);
            state = state.getChildNode(name);
            if (!state.exists()) {
                return Optional.of(p);
            }
        }
        return Optional.empty();
    }
```

mreutegg replied:
Hmm, I can see the benefit of Optional, but in this case the logic is pretty simple and I rather keep it as is.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        @Override
+        public Void call() throws Exception {
```
```suggestion
        public void run() {
        DocumentNodeState state = doc.getNodeAtRevision(ns, headRevision, null);
        if (state != null) {
            Path path = doc.getPath();
            assertAncestorsExist(path.getParent()).ifPresent(missing -> results.put(new OrphanedNode(path, missing, state.getLastRevision())));
        }
    }
```

mreutegg replied:
See my previous comment.

mreutegg committed:
OAK-9918: DocumentNodeStore consistency check

Incorporate review feedback by Rishabh and Julian
Use Oak JsopBuilder to construct JSON

mreutegg committed:
OAK-9918: DocumentNodeStore consistency check

Add documentation for DocumentProcessor
